defaultclient.go

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) 2021 AccelByte Inc. All Rights Reserved.
+// This is licensed software from AccelByte Inc, for limitations
+// and restrictions contact your company contract manager.
+
+package legal
+
+import (
+	"github.com/pkg/errors"
+	"net/http"
+	"time"
+
+	"github.com/AccelByte/iam-go-sdk"
+	"github.com/patrickmn/go-cache"
+)
+
+const (
+	crucialPolicyVersionPath = "/public/policies/version/allCrucial"
+	allAffectedClientID      = "all"
+
+	defaultPolicyVersionCacheTime = 60 * time.Second
+	maxBackOffTime                = 60 * time.Second
+)
+
+type LegalConfig struct {
+	LegalBaseURL                 string
+	PublisherNamespace           string
+	PolicyVersionRefreshInterval time.Duration
+	Debug                        bool
+}
+
+type DefaultLegalClient struct {
+	legalConfig               *LegalConfig
+	policyVersion             map[string][]PolicyVersion
+	policyVersionCache        *cache.Cache
+	policyVersionRefreshError error
+	remotePolicyValidation    func(listPolicyVersion []string, clientID, country, namespace string) (bool, error)
+	// for mocking the HTTP call
+	httpClient HTTPClient
+}
+
+var debug bool
+
+type HTTPClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+// NewDefaultLegalClient creates new Legal DefaultClient
+func NewDefaultLegalClient(config *LegalConfig) LegalClient {
+	if config.PolicyVersionRefreshInterval <= 0 {
+		config.PolicyVersionRefreshInterval = defaultPolicyVersionCacheTime
+	}
+
+	client := &DefaultLegalClient{
+		legalConfig: config,
+		policyVersionCache: cache.New(
+			config.PolicyVersionRefreshInterval,
+			2*config.PolicyVersionRefreshInterval,
+		),
+		httpClient: &http.Client{},
+	}
+
+	client.remotePolicyValidation = client.remoteValidatePolicyVersion
+
+	debug = config.Debug
+
+	log("NewDefaultClient: debug enabled")
+
+	return client
+}
+
+func (client *DefaultLegalClient) StartCachingCrucialLegal() error {
+	err := client.getCrucialPolicyVersion()
+	if err != nil {
+		return logAndReturnErr(
+			errors.WithMessage(err, "StartCachingCrucialLegal: unable to get crucial legal"))
+	}
+
+	go client.refreshCrucialPolicyVersion()
+
+	log("StartCachingCrucialLegal: caching crucial legal start")
+
+	return nil
+}
+
+func (client *DefaultLegalClient) ValidatePolicyVersions(claims *iam.JWTClaims) (bool, error) {
+	// Check for affected clientID
+	if cachedCrucialPolicyVersion, found := client.policyVersionCache.Get(claims.ClientID); found {
```

Comment:
Why need to do this? Wouldn't it be the same if we check all the affected clientIDs once?

Reply:
No, it's different. If we include it in All, it will affected to all client while this only for legal docs that specify affected client.

Reply:
All is only for policy version that base policy didn't specify any affected client